src/tensor/cache.rs

```diff
@@ -29,80 +72,181 @@ pub(crate) struct AllocationKey {
+    fn size(&self) -> usize;
+
+    /// Unsafely converts the elements of a contiguous collection type to another type. Note:
+    /// **This function is wildly unsafe**, see implementations for details
+    unsafe fn transmute_elements<T>(self) -> Self::Output<T>;
+
+    /// Uses transmute_elements to convert to an element type with alignment `align` before dropping.
+    unsafe fn drop_with_alignment(self, align: usize) {
+        match align {
+            1 => drop(self.transmute_elements::<u8>()),
+            2 => drop(self.transmute_elements::<u16>()),
+            4 => drop(self.transmute_elements::<u32>()),
+            8 => drop(self.transmute_elements::<u64>()),
+            16 => drop(self.transmute_elements::<u128>()),
+            _ => panic!("Invalid alignment"),
+        }
+    }
+}
+
+/// (Mostly) Safe wrapper around CacheStorage implementers
+#[derive(Debug)]
+struct CacheWrapper<Ptr: CacheStorage> {
+    ptr: Option<Ptr>,
+    alignment: usize,
+    size: usize,
+}
+
+impl<Ptr: CacheStorage> Drop for CacheWrapper<Ptr> {
+    fn drop(&mut self) {
+        if let Some(ptr) = std::mem::take(&mut self.ptr) {
+            unsafe { ptr.drop_with_alignment(self.alignment) }
+        }
+    }
+}
+
+impl<Ptr: CacheStorage> CacheWrapper<Ptr> {
+    fn from_storage<T>(storage: Ptr::Output<T>) -> Self
+    where
+        Ptr::Output<T>: CacheStorage<Output<u8> = Ptr>,
+    {
+        let layout = Layout::new::<T>();
+        Self {
+            ptr: Some(unsafe { storage.transmute_elements::<u8>() }),
+            alignment: layout.align(),
+            size: layout.size(),
+        }
+    }
+
+    fn check_key(&self, key: &AllocationKey) {
+        assert_eq!(self.alignment, key.alignment, "Alignment does not match");
+        assert_eq!(self.size, key.size, "Size does not match");
+        // Implicitly assumes that T should not have any padding, but this should always be true of
+        // primitive number types.
+        assert_eq!(
+            key.num_bytes % key.size,
+            0,
+            "Key is invalid or type is padded"
+        );
+    }
+
+    fn size(&self) -> usize {
+        self.ptr.as_ref().unwrap().size()
+    }
+
+    // Safety: Same as slice.align_to, but considered safe internally
+    // Produces storage containing uninitialized values
+    fn into_storage<T>(mut self) -> Ptr::Output<T> {
+        let layout = Layout::new::<T>();
+        assert_eq!(layout.align(), self.alignment);
+        assert_eq!(layout.size(), self.size);
+
+        let ptr = std::mem::take(&mut self.ptr).unwrap();
+        unsafe { ptr.transmute_elements() }
+    }
+}
+
+impl<Ptr: CacheStorage> Default for TensorCache<Ptr> {
     fn default() -> Self {
         Self {
             allocations: Default::default(),
             enabled: RwLock::new(false),
+            drop_queue: Default::default(),
+            size: RwLock::new(0),
+            // TODO: default max size
+            max_size: RwLock::new(1_000_000),
```

Comment:
@coreylowman Thoughts on a default `max_size`? I think that 1gb is reasonable in most cases, but maybe we can do something like a percentage of available system memory or we could require the user to specify this value upon enabling the cache.

Reply:
I've decided to require users to specify the maximum size of the cache when they enable it.

Reply:
Thoughts on using an enum to represent different options for this? I'm definitely going to forget what a plain usize would represent
```rust
dev.enable_cache(CacheSize::Unlimited)
dev.enable_cache(CacheSize::NumItems(1000))
dev.enable_cache(CacheSize::Bytes(10))
dev.enable_cache(CacheSize::MB(10))
dev.enable_cache(CacheSize::GB(10))
```

Reply:
Optional num bytes? `Option<usize>`